test/core/event_engine/cf/cf_engine_test.cc

```diff
@@ -80,6 +82,211 @@ TEST(CFEventEngineTest, TestConnectionCancelled) {
   client_signal.WaitForNotification();
 }
 
+namespace {
+std::vector<std::string> ResolvedAddressesToStrings(
+    const std::vector<EventEngine::ResolvedAddress> addresses) {
+  std::vector<std::string> ip_strings;
+  std::transform(addresses.cbegin(), addresses.cend(),
+                 std::back_inserter(ip_strings), [](auto const& address) {
+                   return ResolvedAddressToString(address).value_or("ERROR");
+                 });
+  return ip_strings;
+}
+}  // namespace
+
+TEST(CFEventEngineTest, CFEventEngineTest_TestCreateDNSResolver_Test) {
+  grpc_core::MemoryQuota memory_quota("cf_engine_test");
+  auto cf_engine = std::make_shared<CFEventEngine>();
+
+  EXPECT_TRUE(cf_engine->GetDNSResolver({}).status().ok());
+  EXPECT_TRUE(cf_engine->GetDNSResolver({.dns_server = ""}).status().ok());
+  EXPECT_EQ(
+      cf_engine->GetDNSResolver({.dns_server = "8.8.8.8"}).status().code(),
+      absl::StatusCode::kInvalidArgument);
+  EXPECT_EQ(
+      cf_engine->GetDNSResolver({.dns_server = "8.8.8.8:53"}).status().code(),
+      absl::StatusCode::kInvalidArgument);
+  EXPECT_EQ(
+      cf_engine->GetDNSResolver({.dns_server = "invalid"}).status().code(),
+      absl::StatusCode::kInvalidArgument);
+}
+
+TEST(CFEventEngineTest, TestResolveLocalhost) {
+  grpc_core::Notification resolve_signal;
+
+  auto cf_engine = std::make_shared<CFEventEngine>();
+  auto dns_resolver = cf_engine->GetDNSResolver({});
+
+  dns_resolver.value()->LookupHostname(
+      [&resolve_signal](auto result) {
+        EXPECT_TRUE(result.status().ok());
+        EXPECT_THAT(ResolvedAddressesToStrings(result.value()),
+                    testing::UnorderedElementsAre("127.0.0.1:80", "[::1]:80"));
+
+        resolve_signal.Notify();
+      },
+      "localhost", "80");
+
+  resolve_signal.WaitForNotification();
+}
+
+TEST(CFEventEngineTest, TestResolveRemote) {
+  grpc_core::Notification resolve_signal;
+
+  auto cf_engine = std::make_shared<CFEventEngine>();
+  auto dns_resolver = cf_engine->GetDNSResolver({});
+
+  dns_resolver.value()->LookupHostname(
+      [&resolve_signal](auto result) {
+        EXPECT_TRUE(result.status().ok());
+        EXPECT_THAT(ResolvedAddressesToStrings(result.value()),
+                    testing::UnorderedElementsAre("127.0.0.1:80", "[::1]:80"));
+
+        resolve_signal.Notify();
+      },
+      "localtest.me:80", "443");
```

Comment:
Word of caution that any test that relies on an external service may A) become flaky even though the implementation is fine, and B) may be pretty unfriendly if we run `bazel test --runs_per_test=1000 ...`.

Reply:
Yeah, but specifying a local dns server is not supported for this implementation.
I guess it's less likely having a DNS issue comparing to other external services (homebrew for instance) we are using to setup the environment.
We used it for [iOS PerfTest](https://github.com/grpc/grpc/blob/v1.56.0/src/objective-c/tests/PerfTests/PerfTests.m#L363) for a while which was pretty stable.
So I suppose it's better have test coverage? Or are there any alternative Google services?

Stress runs should be fine as I imagine it'll hit the cache.

Reply:
PS: I tested with domains with multiple records (e.g. wsj.com) and it works. But I don't know a service that can always response the same set of IPs, so I don't have a test case about it.

Reply:
I don't have a better solution for iOS, I'm not sure what options you have in your test environments. Maybe it will be fine. I'm simply noting that if there is noise in the tests because of services you don't control, those tests might end up not being very valuable.

> I tested with domains with multiple records (e.g. wsj.com) and it works (got all 4 v4 IPs). 

Great to hear!